test/unit/math/mix/fun/complex_schur_decompose_test.cpp

```diff
@@ -0,0 +1,101 @@
+    stan::test::expect_ad(f, x);
+  }
+
+  Eigen::MatrixXd a32(3, 2);
+  a32 << 3, -5, 7, -7.2, 9.1, -6.3;
+  EXPECT_THROW(f(a32), std::invalid_argument);
+}
+
+template <typename V>
+void test_complex_schur_decompose(const Eigen::MatrixXd& x) {
+  using stan::math::complex_schur_decompose_t;
+  using stan::math::complex_schur_decompose_u;
+  using stan::math::value_of_rec;
+  Eigen::Matrix<V, -1, -1> X(x.rows(), x.cols());
+  for (int i = 0; i < x.size(); ++i)
+    X(i) = x(i);
+  auto T = complex_schur_decompose_t(X);
+  auto U = complex_schur_decompose_u(X);
+  auto X2 = U * T * U.adjoint();
+  for (int j = 0; j < x.cols(); ++j) {
+    for (int i = 0; i < x.rows(); ++i) {
+      EXPECT_FLOAT_EQ(x(i, j), value_of_rec(X2(i, j).real()));
+    }
+  }
+}
+
+// TODO(carpenter): complex_schur_decompose only works for real types,
+// returning complex types;  the functions work for double types, but
+// not autodiff complex types;  what's missing is compound arithmetic
+// for matrix base type lhs and complex rhs
+
+template <typename V>
+void test_complex_schur_decompose_complex(const Eigen::MatrixXd& x) {
+  using stan::math::complex_schur_decompose_t;
+  using stan::math::complex_schur_decompose_u;
+  using stan::math::value_of_rec;
+  Eigen::Matrix<std::complex<V>, -1, -1> X(x.rows(), x.cols());
+  for (int i = 0; i < x.size(); ++i)
+    X(i) = std::complex<double>(x(i), i);
+  auto T = complex_schur_decompose_t(X);
+  auto U = complex_schur_decompose_u(X);
+  auto X2 = U * T * U.adjoint();
+  for (int j = 0; j < x.cols(); ++j) {
+    for (int i = 0; i < x.rows(); ++i) {
+      EXPECT_NEAR(value_of_rec(X(i, j).real()), value_of_rec(X2(i, j).real()),
+                  1e-8);
+      EXPECT_NEAR(value_of_rec(X(i, j)).imag(), value_of_rec(X2(i, j).imag()),
+                  1e-8);
+    }
+  }
+}
+
+TEST(mathMixFun, complexSchurDecompose) {
+  using d_t = double;
+  using v_t = stan::math::var;
+  using fd_t = stan::math::fvar<d_t>;
+  using ffd_t = stan::math::fvar<fd_t>;
+  using fv_t = stan::math::fvar<v_t>;
+  using ffv_t = stan::math::fvar<fv_t>;
+  for (const auto& x : stan::test::square_test_matrices(1, 3)) {
+    test_complex_schur_decompose<d_t>(x);
+    test_complex_schur_decompose<v_t>(x);
+    test_complex_schur_decompose<fd_t>(x);
+    test_complex_schur_decompose<ffd_t>(x);
+    test_complex_schur_decompose<fv_t>(x);
+    test_complex_schur_decompose<ffv_t>(x);
+  }
+  for (const auto& x : stan::test::square_test_matrices(1, 3)) {
+    test_complex_schur_decompose_complex<d_t>(x);
+    // TODO(carpenter): // require expanding compound arithmetic
+    //                  // to more general Eigen types for *=, -=, etc.
+    // test_complex_schur_decompose_complex<v_t>(x);
+    // test_complex_schur_decompose_complex<fd_t>(x);
+    // test_complex_schur_decompose_complex<ffd_t>(x);
+    // test_complex_schur_decompose_complex<fv_t>(x);
+    // test_complex_schur_decompose_complex<ffv_t>(x);
```

Comment:
Do we want to wait on schur until this todo is resolved?

Reply:
I meant to say that `test_schur_decompose` only works for `double`.   As far as I can tell, the Schur decompositions themselves work as intended.

The problem is that we can't put them back together because of missing operations.  It's line 42 that's the problem:

```
 auto X2 = U * T * U.adjoint();
```

I'm very unclear on why this doesn't work given that all the other complex arithemtic works.  I think it has to do with the expression templates not picking up all the right traits.  Which makes me wonder how the actual calls to Schur and eigendecomposition and what-not work.

Reply:
Let me take a stab at it this weekend for a little bit. If I can't figure it out I think we should just remove it

Reply:
This may be a bigger issue, I'm seeing some things like

lib/eigen_3.3.3/Eigen/src/Core/GeneralProduct.h:282:25: error: no match for ‘operator+=’
 (operand types are ‘Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<
std::complex<stan::math::var>, -1, 1, 0, -1, 1>, 0, Eigen::Stride<0, 0> > >’ 
and ‘std::complex<stan::math::var>’)
  282 |           dest.matrix() += actualAlpha * MappedDest(actualDestPtr, dest.size());
Do you think it's because Eigen can't find the += for complex?

Reply:
I think it is a bigger issue and I think it's a problem with assignment.  I'm going to try to get to the bottom of it.  I'm surprised all these algorithms that work in the complex domain work.  But then Schur broke with complex inputs, so maybe there are different implementation paths.